src/dbnode/network/server/tchannelthrift/cluster/service.go

```diff
@@ -43,56 +45,73 @@ type service struct {
 	sync.RWMutex
 
 	client client.Client
-	active client.Session
+	active map[sessionOpts]client.Session
 	opts   client.Options
 	idPool ident.Pool
 	health *rpc.HealthResult_
 }
 
+// a composite key to lookup the type of session for the request.
+type sessionOpts struct {
+	readConsistency        topology.ReadConsistencyLevel
+	equalTimestampStrategy encoding.IterateEqualTimestampStrategy
+}
+
 // NewService creates a new cluster TChannel Thrift service
-func NewService(client client.Client) rpc.TChanCluster {
+func NewService(c client.Client) rpc.TChanCluster {
 	s := &service{
-		client: client,
-		opts:   client.Options(),
-		idPool: client.Options().IdentifierPool(),
+		client: c,
+		active: make(map[sessionOpts]client.Session),
+		opts:   c.Options(),
+		idPool: c.Options().IdentifierPool(),
 		health: &rpc.HealthResult_{Ok: true, Status: "up"},
 	}
 	return s
 }
 
 func (s *service) session() (client.Session, error) {
+	return s.sessionForOpts(s.sessionOptsFromClusterQueryOpts(nil))
+}
+
+func (s *service) sessionForOpts(opts sessionOpts) (client.Session, error) {
 	s.RLock()
-	session := s.active
+	session := s.active[opts]
 	s.RUnlock()
 	if session != nil {
 		return session, nil
 	}
 
 	s.Lock()
-	if s.active != nil {
-		session := s.active
+	session = s.active[opts]
+	if session != nil {
 		s.Unlock()
 		return session, nil
 	}
-	session, err := s.client.DefaultSession()
```

Comment:
it's fine we're not using the default session anymore, because this code is already reusing sessions.